components/brave_wallet/browser/fil_tx_manager.cc

```diff
@@ -29,25 +42,160 @@ FilTxManager::FilTxManager(TxService* tx_service,
+  const std::string gas_premium = tx->gas_premium();
+  auto gas_limit = tx->gas_limit();
+  if (!gas_limit || gas_fee_cap.empty() || gas_premium.empty()) {
+    GetEstimatedGas(from, std::move(tx_ptr), std::move(callback));
+  } else {
+    ContinueAddUnapprovedTransaction(
+        from, std::move(tx_ptr), std::move(callback), gas_premium, gas_fee_cap,
+        gas_limit, mojom::FilecoinProviderError::kSuccess, "");
+  }
+}
+
+void FilTxManager::GetEstimatedGas(const std::string& from,
+                                   std::unique_ptr<FilTransaction> tx,
+                                   AddUnapprovedTransactionCallback callback) {
+  const std::string gas_premium = tx->gas_premium();
+  const std::string gas_fee_cap = tx->gas_fee_cap();
+  auto gas_limit = tx->gas_limit();
+  uint64_t nonce = tx->nonce() ? tx->nonce().value() : 0;
+  const std::string value = tx->value();
+  const std::string max_fee = tx->max_fee();
+  auto to = tx->to().EncodeAsString();
+  json_rpc_service_->GetFilEstimateGas(
+      from, to, gas_premium, gas_fee_cap, gas_limit, nonce, max_fee, value,
+      base::BindOnce(&FilTxManager::ContinueAddUnapprovedTransaction,
+                     weak_factory_.GetWeakPtr(), from, std::move(tx),
+                     std::move(callback)));
+}
+
+void FilTxManager::ContinueAddUnapprovedTransaction(
+    const std::string& from,
+    std::unique_ptr<FilTransaction> tx,
+    AddUnapprovedTransactionCallback callback,
+    const std::string& gas_premium,
+    const std::string& gas_fee_cap,
+    int64_t gas_limit,
+    mojom::FilecoinProviderError error,
+    const std::string& error_message) {
+  tx->set_gas_premium(gas_premium);
+  tx->set_fee_cap(gas_fee_cap);
+  tx->set_gas_limit(gas_limit);
+  FilTxMeta meta(std::move(tx));
+  meta.set_id(TxMeta::GenerateMetaID());
+  meta.set_from(FilAddress::FromAddress(from).EncodeAsString());
+  meta.set_created_time(base::Time::Now());
+  meta.set_status(mojom::TransactionStatus::Unapproved);
+  GetFilTxStateManager()->AddOrUpdateTx(meta);
+  std::move(callback).Run(true, meta.id(), "");
+}
+
+void FilTxManager::AddUnapprovedTransaction(
+    mojom::TxDataUnionPtr tx_data_union,
+    const std::string& from,
+    AddUnapprovedTransactionCallback callback) {
+  AddUnapprovedTransaction(std::move(tx_data_union->get_fil_tx_data()), from,
+                           std::move(callback));
 }
 
 void FilTxManager::ApproveTransaction(const std::string& tx_meta_id,
-                                      ApproveTransactionCallback) {
-  NOTIMPLEMENTED();
+                                      ApproveTransactionCallback callback) {
+  std::unique_ptr<FilTxMeta> meta =
+      GetFilTxStateManager()->GetFilTx(tx_meta_id);
+  if (!meta) {
+    DCHECK(false) << "Transaction should be found";
+    std::move(callback).Run(
+        false,
+        mojom::ProviderErrorUnion::NewFilecoinProviderError(
+            mojom::FilecoinProviderError::kActorNotFound),
```

Comment:
What does ActorNotFound mean in filecoin?

Reply:
that actor not found to process specific request, transaction in this case

Reply:
I wonder if we should just follow other coin type to use kInternalError and IDS_BRAVE_WALLET_TRANSACTION_NOT_FOUND here? At least the message seems more clear to me.

Reply:
done